app/assets/javascripts/discourse/app/controllers/topic.js

```diff
@@ -1292,6 +1294,42 @@ export default Controller.extend(bufferedProperty("model"), {
     this.model.destroy(this.currentUser);
   },
 
+  retryOnRateLimit(times, promise, topicId) {
+    const currentTopicId = this.get("model.id");
+    topicId = topicId || currentTopicId;
+    if (topicId !== currentTopicId || times <= 0) {
+      // we navigated to another topic, so skip
+      return;
+    }
+
+    if (this.retryRateLimited) {
+      return;
+    }
+
+    promise().catch((e) => {
+      const xhr = e.jqXHR;
+      if (
+        xhr &&
+        xhr.status === 429 &&
+        xhr.responseJSON &&
+        xhr.responseJSON.extras &&
+        xhr.responseJSON.extras.wait_seconds
+      ) {
+        let waitSeconds = xhr.responseJSON.extras.wait_seconds;
+        if (waitSeconds < 5) {
+          waitSeconds = 5;
```

Comment:
Checking the rest of the code.. there is server randomization, uniform between 5-10sec @ 10ms.

The client is in a better position to implement randomized exponential backoff because it has better visibility into how many consecutive attempts it's made. Uniform randomization is decent and decays into a normal distribution after multiple attempts.

Checking other implementations.. they trust the backoff value from the server, and use randomized exponential when the server can't give a backoff value.

https://github.com/menghanl/grpc-go/blob/4e06cefbe5f1619362397ea2ad765f1c6311f748/stream.go#L546-L558

Okay, so concrete suggestion: Please document that the value from the server is randomized @ 10ms granularity. Don't floor at 5 seconds, floor at 0.5 seconds - self documenting + allows more server customization flexiblity.

Reply:
I do not really agree with 0.5 seconds here as a minimum. Certainly without seeing longer delays as a problem in production. End user will be oblivious here, it will look like a reply happened just now even if it happened 5 seconds ago.

I really need to observe logs from an event with these params prior to tweaking anything.